# 第十二章: 异常处理 #
在程序设计和运行的过程中，发生错误是不可避免的。尽管Java语言的设计从根本上提供了便于写出整洁、安全的代码的方法，并且程序员也尽量地减少错误的产生，但使程序被迫停止的错误的存在仍然不可避免。为此，Java提供了异常处理机制来帮助程序员检查可能出现的错误，保证了程序的可读性和可维护性。Java中将异常封装到一个类中，出现错误时，就会抛出一个异常。

## 本章目标：
- 了解异常的概念
- 掌握捕获异常
- 了解Java中常见的异常
- 掌握自定义异常
- 了解如何在方法中抛出异常
- 了解运行时异常的种类
- 了解异常处理的使用原则

## 一、异常概述 ##
在程序中，错误可能产生于程序员没有预料到的各种情况，或者是超出了程序员可控范围的环境因素，如用户的坏数据、试图打开一个根本不存在的文件等。在Java中这种在程序运行时可能出现的一些错误称为异常。异常是一个在程序执行期间发生的事件，它中断了正在执行的程序的正常指令流。

实例：在项目中创建类Baulk，在主方法中定义int型变量，将0作为除数赋值给该变量。

	...
	
程序运行的记过报告发生了算术异常 ArithmeticException，系统不再执行下去，提前结束。这种情况就是所说的异常。

Java语言是一门面向对象的编程语言，因此，异常在Java语言中也是作为类的实例的形式出现的。当某一方法中发生错误时，这个方法会创建一个对象，并且把它传递给正在运行的系统。这个对象就是异常对象。通过异常处理机制，可以将非正常情况下的处理代码与程序的主逻辑分离，即在编写程序主流程的同时在其他地方处理异常。

## 二、处理程序异常错误 ##
为了保证程序有效地执行，需要对发生的异常进行相应的处理。在Java中，如果某个方法抛出异常，既可以在当前方法中进行捕捉，然后处理该异常，也可以将异常向上抛出，由方法调用者来处理。

### 2.1 错误 ###
异常产生后，如果不做任何处理，程序就会被终止。

实例：在项目中创建类Thundering，在主方法中实现将非字符型数值转换为int型。运行程序，系统会报出异常提示。

	...
	
运行结果图：

### 2.2 捕获异常 ###
Java语言的异常捕获结构由try, catch和finally三部分组成。其中，try语句块存放的是可能发生异常的Java语句；catch程序块在try语句块之后，用来激发被捕获的异常；finally语句块是异常处理结构的最后执行部分，无论try语句块中的代码如何退出，都将执行finally语句块。

语法格式：

	try{
		// 程序代码块
	}
	catch(Exceptiontype1 e){
		// 对Exceptiontype1的处理
	}
	catch(Exceptiontype2 e){
		// 对Exceptiontype2的处理
	}
	...
	finally{
		// 程序块
	}
	
通过异常处理器的语法可知，异常处理器大致分为try-catch语句块和finally语句块。

实例：在项目中创建Take，在主方法中使用try-catch语句块将可能出现的异常语句进行异常处理。

	...
	
运行结果图：

从图中可以看出，程序仍然输出最后的提示信息，没有因为异常而终止。将可能出现异常的代码用try-catch语句块进行处理，当try代码块中的语句发生异常时，程序就会调转到catch代码块中执行，执行完catch代码块中的程序代码后，将继续执行catch代码块后的其他代码，而不会执行try代码快中发生异常语句后面的代码。由此可见，Java的异常处理是结构化的，不会因为一个异常影响整个程序的执行。


> 注意：Exception是try代码块传递给catch代码块的变量类型，e是变量名。catch代码块中语句“e.getMessage();”用于输出错误性质。通常，异常处理常用以下3个函数获取异常的有关信息。

- getMessage()函数：输出错误性质。
- toString()函数：给出异常的类型与性质。
- printStackTrace()函数：指出异常的类型、性质、栈层次及出现在程序中的位置。

#### finally语句块 ####
完整的异常处理语句一定要包含finally语句，无论程序中有无异常发生，并且无论之间的try-catch是否顺利执行完毕，都会执行finally语句。

在以下4种特殊情况下，finally块不会被执行：

- 在finally语句块中发生了异常。
- 在前面的代码中使用了 System.exit() 退出程序。
- 程序所在的线程死亡。
- 关闭CPU。

## 三、Java常见异常 ##
在Java中提供了一些异常用来描述经常发生的错误，其中，有的需要程序员进行捕获处理或声明抛出，有的是由Java虚拟机自动进行捕获处理的。Java中常见的异常类如下所示：

**api:** [https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html](https://docs.oracle.com/javase/8/docs/api/java/lang/Exception.html)


	Direct Known Subclasses:
	
	AclNotFoundException,
	ActivationException,
	AlreadyBoundException,
	ApplicationException,
	AWTException,
	BackingStoreException,
	BadAttributeValueExpException,
	BadBinaryOpValueExpException,
	BadLocationException,
	BadStringOperationException,
	BrokenBarrierException,
	CertificateException,
	CloneNotSupportedException,
	DataFormatException,
	DatatypeConfigurationException,
	DestroyFailedException,
	ExecutionException,
	ExpandVetoException,
	FontFormatException,
	GeneralSecurityException,
	GSSException,
	IllegalClassFormatException,
	InterruptedException,
	IntrospectionException,
	InvalidApplicationException,
	InvalidMidiDataException,
	InvalidPreferencesFormatException,
	InvalidTargetObjectTypeException,
	IOException,
	JAXBException,
	JMException,
	KeySelectorException,
	LambdaConversionException,
	LastOwnerException,
	LineUnavailableException,
	MarshalException,
	MidiUnavailableException,
	MimeTypeParseException,
	MimeTypeParseException,
	NamingException,
	NoninvertibleTransformException,
	NotBoundException,
	NotOwnerException,
	ParseException,
	ParserConfigurationException,
	PrinterException,
	PrintException,
	PrivilegedActionException,
	PropertyVetoException,
	ReflectiveOperationException,
	RefreshFailedException,
	RemarshalException,
	RuntimeException,
	SAXException,
	ScriptException,
	ServerNotActiveException,
	SOAPException,
	SQLException,
	TimeoutException,
	TooManyListenersException,
	TransformerException,
	TransformException,
	UnmodifiableClassException,
	UnsupportedAudioFileException,
	UnsupportedCallbackException,
	UnsupportedFlavorException,
	UnsupportedLookAndFeelException,
	URIReferenceException,
	URISyntaxException,
	UserException,
	XAException,
	XMLParseException,
	XMLSignatureException,
	XMLStreamException,
	XPathException

## 四、自定义异常 ##
使用Java内置的异常类可以描述在编程时出现的大部分异常情况。除此之外，用户只需继承Exception类即可自定义异常类。

在程序中使用自定义异常类，大体可分为以下几个步骤：

- 创建自定义异常类。
- 在方法中通过throw关键字抛出异常对象。
- 如果在当前抛出异常的方法中处理异常，可以使用try-catch语句块捕获并处理，否则在方法的声明处通过throws关键字指明要抛出给方法调用者的异常，继续进行下一步操作。
- 在出现异常方法的调用者中捕获并处理异常。

## 五、在方法中抛出异常 ##
- throws
- throw

若某个方法可能会发生异常，但不想在当前方法中处理这个异常，则可以使用throws, throw关键字在方法中抛出异常。

### 5.1 使用throws关键字抛出异常 ###
throws关键字通常被应用在声明方法时，用来指定方法可能抛出的异常。多个异常可使用逗号分隔。

实例：在项目中创建类Shoot，在该类中创建方法pop()，在该方法中抛出NegativeArraySizeException异常，在主方法中调用该方法，并实现异常处理。

	...
	



### 5.2 使用throw关键字抛出异常 ###

## 六、运行时异常 ##

## 七、异常使用原则 ##